design/src/rfcs/rfc0022_refine_builder.md

```diff
@@ -0,0 +1,556 @@
+
+```rust
+impl PokemonService<()> {
+    /// Constructs a builder for [`PokemonService`].
+    pub fn builder<Body, Plugin>(plugin: Plugin) -> PokemonServiceBuilder<Body, Plugin> {
+        PokemonServiceBuilder {
+            check_health: None,
+            do_nothing: None,
+            get_pokemon_species: None,
+            get_server_statistics: None,
+            capture_pokemon: None,
+            get_storage: None,
+            plugin,
+        }
+    }
+}
+```
+
+In practice, this is unlikely to have any impact on developers using `smithy-rs`.  
+We have seen how cumbersome it is to break the startup logic into different functions using the current service builder API. Developers are most likely specifying all plugins and routes in the same function even if the current API allows them to intersperse route registrations and plugin registrations: they would simply have to re-order their registration statements to adopt the API proposed in this RFC.
+
+### Alternatives: boxed trait objects 
+
+Trait objects are another common approach to achieve type erasure: instead of storing the raw handler in `PokemonServiceBuilder`, we box it and cast it to a `dyn Upgradable<...>`.  
+Using trait objects would allow us to delay upgrading to `Route` until `build` is called, mimicking the behaviour of the current API and allowing developers to register plugins at any point before calling `build`.  
+Using trait objects would unfortunately have a non-zero runtime impact: we would be introducing an extra layer of indirection (double-boxing the handler), affecting the performance of our application at runtime.  
+Given the above, we do not suggest we fall back on trait objects to improve the ergonomics of our API.
+
+### Builder extensions: what now?
+
+The `Pluggable` trait was an interesting development out of [RFC 20]: it allows you to attach methods to a service builder using an extension trait.
+
+```rust
+/// An extension to service builders to add the `print()` function.
+pub trait PrintExt: aws_smithy_http_server::plugin::Pluggable<PrintPlugin> {
+    /// Causes all operations to print the operation name when called.
+    ///
+    /// This works by applying the [`PrintPlugin`].
+    fn print(self) -> Self::Output
+        where
+            Self: Sized,
+    {
+        self.apply(PrintPlugin)
+    }
+}
+```
+
+This pattern needs to be revisited if we want to move forward with this RFC, since new plugins cannot be registered after the builder has been instantiated.  
+My recommendation would be to implement `Pluggable` for `PluginStack`, providing the same pattern ahead of the creation of the builder:
+
+```rust
+// Currently you'd have to go for `PluginStack::new(IdentityPlugin, IdentityPlugin)`,
+// but that can be smoothed out even if this RFC isn't approved.
+let plugin_stack = PluginStack::default()
```

Comment:
I should have called this out in the RFC, but I am indeed not convinced this is the right way forward for the same reason you mentioned - it can be confusing and lead to surprising runtime behaviour.
Especially considering how existing web frameworks in Rust ignore _at what point_ a middleware is registered and apply it to all routes at the same scope/level.

Reply:
If the intention is that the plugin would only affect the operations registered so far, I think it's ok.

We can also consider providing a way to apply plugins (i.e. layers to the operation before it's HTTP-upgraded) more fine-grained, to _specific_ operations via a method on `Operation`, instead of the current way in #1837. But this is a one-way door, we can add this functionality later if there's user interest.

Reply:
> We can also consider providing a way to apply plugins (i.e. layers to the operation before it's HTTP-upgraded) more fine-grained, to specific operations via a method on Operation, instead of the current way in https://github.com/awslabs/smithy-rs/pull/1837.

I think just adding a blanket impl

```rust
impl<P, Op, Pl, S, L> Pluggable<Pl> for Operation<S, L> where Pl: Plugin<P, Op, S, L> {
    type Output = Operation<Pl::Service, Pl::Layer>;

   fn apply(self, new_plugin: Pl) -> Self::Output {
       new_plugin.map(self)
   }
}
```

to the `aws-smithy-http-server`

would allow you to

```rust
op.print()
```

Reply:
I have captured both as an alternative in https://github.com/awslabs/smithy-rs/pull/1859/commits/90ae7aa8a02b3e39b767c8e9379e244d47b7dd98
I've added as well a sketch of an API for attaching plugins to groups of routes using a nested/scoped builder.